Refactor RecordFunction to use CustomClass

cloudhan opened the pull request:
Fixes #35026.

There used to be a try in #35596 but is long left behind and then abandoned. This PR redo it on relative new codebase.

And #62390 is related, this PR should remove additional `zeros` and `empty` op from profiling trace. Whether there will be a performance improvement is to be measured.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @ilia-cher @robieta @chaekit @gdankel @bitfort @ngimel @orionr @nbcsm @guotuofeng @guyang3532 @gaoteng-git

cloudhan committed:
Make `RecordFunction` stops using `cpp_custom_type_hack`

cloudhan committed:
remove deprecation warning mute logic from cmake

cloudhan commented:
Open this PR to see what will the CI says. Locally, all tests in `test/test_profiler.py` passed.

cloudhan committed:
update `record_function` context manager for `jit.script`

cloudhan committed:
fix some test in rpc_test.py

cloudhan commented:
After move to CustomClass, this line 
https://github.com/pytorch/pytorch/blob/773a8eede437de972311b21824694e4fb120f6a2/torch/autograd/profiler_legacy.py#L186 
is causing one of `rpc_async_jit#__torch__.script_add_ones(worker0 -> worker0)` being filtered out. Then test failure:

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 418, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 644, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 689, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 543, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 420, in wrapper
    fn()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/dist_utils.py", line 73, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/jit/rpc_test.py", line 1229, in test_record_function_on_caller_rpc_async
    self.assertEqual(2, len(jit_rpc_events))
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1630, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
  File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : Scalars failed to compare as equal! 2 != 1
```
see in [log](https://app.circleci.com/pipelines/github/pytorch/pytorch/359973/workflows/bdad8755-dd3c-4bf8-9196-4d9d747a264d/jobs/15184038)

Just a note before push.

cloudhan committed:
use `start_us` instead of `node_id` to workaround test failure due to filtering

cloudhan committed:
mute mypy and leave some note

cloudhan commented:
@gdankel This PR breaks the back compatibility due to refactoring, would you please confirm on it or find someone is responsible for it to do a double check?

cloudhan committed:
remove torch header from record_function

cloudhan commented:
Backward compatibility breakage is expected and intended.

ngimel commented:
`_record_function_exit` is not part of public API (starts with underscore) so that's fine.

ngimel commented:
Please fix bc test by adding this function to the exception list with a future date. 
Can someone from the profiler team please review and land? cc @chaekit, @gdankel

cloudhan commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -16,40 +15,31 @@ namespace profiler {
-at::Tensor record_function_enter(const std::string& name) {
-  auto rec = std::make_unique<at::RecordFunction>(at::RecordScope::USER_SCOPE);
+c10::intrusive_ptr<at::RecordFunction> record_function_enter(const std::string& name) {
+  auto rec = c10::make_intrusive<at::RecordFunction>(at::RecordScope::USER_SCOPE);
```
because `RecordFunction` is native pytorch objects now, that is, intrusive_ptr refcounted.

cloudhan commented on torch/autograd/profiler_legacy.py:
```diff
@@ -186,7 +186,7 @@ def _get_record_key(record):
                         prev_record.name() == record.name()
                         and prev_record.kind() == record.kind()
-                        and prev_record.node_id() == record.node_id()
+                        and prev_record.start_us() == record.start_us()
```
See this comment. https://github.com/pytorch/pytorch/pull/62710#issuecomment-893203087

cloudhan replied:
This testing failure occurs in this branch right after Ilia's refactor. It serve as quick dirty fix toward legacy code. If you insist, I can do more investigation about it.

ngimel replied:
YEs please, relying on the time stamp to identify duplicates can result in false positive matches.

ngimel replied:
@ilia-cher do you know why your refactor would cause this?

cloudhan replied:
@ngimel 

https://github.com/pytorch/pytorch/blob/0507c316cd63cecd899849337d56fc38a8f62c2b/torch/autograd/profiler_legacy.py#L175

after this line, I added a print 
```py
print("node_id:", record.node_id(), "start_us:", record.start_us(), "name:", record.name())
```

trigger the test simply with 
```
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_record_function_on_caller_rpc_async
```

<details>
  <summary>expand to see the stdout</summary>

```
node_id: 0 start_us: 1629782540852445.2 name: aten::add
node_id: 0 start_us: 1629782540852593.8 name: 
node_id: 0 start_us: 1629782540856163.8 name: __exit__
node_id: 0 start_us: 1629782540856210.5 name: 
node_id: 0 start_us: 1629782540691356.5 name: __start_profile
node_id: 0 start_us: 1629782540696474.5 name: defaults
node_id: 0 start_us: 1629782540696482.8 name: 
node_id: 0 start_us: 1629782540718015.0 name: defaults
node_id: 0 start_us: 1629782540718018.0 name: 
node_id: 0 start_us: 1629782540720749.8 name: defaults
node_id: 0 start_us: 1629782540720753.0 name: 
node_id: 0 start_us: 1629782540746427.5 name: defaults
node_id: 0 start_us: 1629782540746430.5 name: 
node_id: 0 start_us: 1629782540748283.5 name: defaults
node_id: 0 start_us: 1629782540748286.5 name: 
node_id: 0 start_us: 1629782540754889.5 name: defaults
node_id: 0 start_us: 1629782540754892.5 name: 
node_id: 0 start_us: 1629782540838905.8 name: record_function_on_caller_rpc_async
node_id: 0 start_us: 1629782540838916.5 name: aten::ones
node_id: 0 start_us: 1629782540838946.8 name: aten::empty
node_id: 0 start_us: 1629782540838961.8 name: 
node_id: 0 start_us: 1629782540838975.2 name: aten::fill_
node_id: 0 start_us: 1629782540839018.5 name: 
node_id: 0 start_us: 1629782540839022.2 name: 
node_id: 0 start_us: 1629782540839775.0 name: __enter__
node_id: 0 start_us: 1629782540839782.8 name: foo
node_id: 0 start_us: 1629782540839795.0 name: 
node_id: 0 start_us: 1629782540839840.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 start_us: 1629782540840022.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 start_us: 1629782540850260.0 name: 
node_id: 0 start_us: 1629782540851111.2 name: 
node_id: 0 start_us: 1629782540856195.8 name: 
node_id: 0 start_us: 1629782540856257.5 name: 
node_id: 0 start_us: 1629782540856288.5 name: __stop_profile
node_id: 1 start_us: 1629782540840774.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __start_profile
node_id: 1 start_us: 1629782540849054.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: script_add_ones
node_id: 1 start_us: 1629782540849112.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::add
node_id: 1 start_us: 1629782540849179.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849251.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849292.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __stop_profile
node_id: 1 start_us: 1629782540840564.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __start_profile
node_id: 1 start_us: 1629782540842807.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: defaults
node_id: 1 start_us: 1629782540842812.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846464.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::ones
node_id: 1 start_us: 1629782540846489.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::empty
node_id: 1 start_us: 1629782540846500.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846510.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::fill_
node_id: 1 start_us: 1629782540846545.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846548.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849036.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: script_add_ones
node_id: 1 start_us: 1629782540849077.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::add
node_id: 1 start_us: 1629782540849176.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849231.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849271.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __stop_profile
```
</details>

as you can see, most record has duplicated `node_id`...

cloudhan replied:
OK, it seems that I have some misunderstanding of `node_id` in this case. It is not the ID of the profiling event. It is the machine node id that emit the event. I will revert the change of this line.

cloudhan replied:
@ngimel OK, I nailed it. I checkout 773a8eede437de972311b21824694e4fb120f6a2, the commit of @ilia-cher 's refactor. With the previous print hack to inspect into the deduplication process

```
...
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zeros
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zero_
node_id: 0 kind: push name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zeros
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zero_
node_id: 0 kind: push name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 kind: push name: aten::empty
...
```

The reason the `...script_add_ones...`s were not deduplicated is because original `cpp_custom_type_hack` introduce `aten::empty`, `aten::zeros`, `aten::empty`, `aten::zero_` events between two `...script_add_ones...`s. Now, without the hack, two `...script_add_ones...`s appears **consecutive**, their `name`, `kind`, `node_id` are the same, and then deduplicated.

cloudhan replied:
@ngimel I update the workaround. Instead of trying to fix the deduplication logic in `profiler_legacy.py`. Instead, I added a dummy operation between the two `rpc_async` in the test case. So that they are not consecutive anymore. How to avoid false-positively classify  consecutive `...script_add_ones...`s as duplicated is left to pytorch team.

rohan-varma commented on torch/autograd/profiler.py:
```diff
@@ -458,15 +459,19 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
         return profiled_future
 
+    @property
+    def handle(self):
```
Do we need to add error checking to ensure this list isn't empty?

cloudhan replied:
The `RecordFunction` object originally hold by this context manager is called `handle`. This property exists is purely due to existing tests use this internal object for testing.

Anyway, if we add error checking, do we simply capture the `IndexError` and reraise it with clearer `RuntimeError`?

cloudhan replied:
I added an error checking with comprehensive `RuntimeError` msg.

rohan-varma commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -67,6 +57,7 @@ c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut(
 
 // Internal only, do not use directly, use Python's record_function()
 TORCH_LIBRARY_FRAGMENT(profiler, m) {
+    m.class_<at::RecordFunction>("RecordFunction");
```
nit: `_RecordFunction` as it's not meant to be used publicly?

cloudhan replied:
Updated.

cloudhan committed:
add _record_function_exit to backward compatibility breakage allowlist

cloudhan committed:
rename RecordFunction to _RecordFunction on python side, since it is for internal usage

cloudhan committed:
Revert "use `start_us` instead of `node_id` to workaround test failure due to filtering"

This reverts commit b837f9c4dd701d3efd4b2a0274afa24cf26d2528.

cloudhan committed:
workaround test issue

Add minor change toward the test case. So that the deduplication in
https://github.com/pytorch/pytorch/pull/62710#discussion_r694680279
won't false-positively deduplicate the two script_add_ones

cloudhan committed:
break long line

cloudhan commented:
### Backward Compatibility Breakage

This PR will break previously `jit.script` generated model if and only if the model itself contains `record_function`.
| `torch.jit.save`-ed file   | Run with PyTorch 1.9.0 | Run with This PR |
| --- | --- | --- |
| PyTorch 1.9.0  jit.script  | Good | Bad |
| PyTorch 1.9.0  jit.trace | Good | Good |
| This PR  jit.script | Bad | Good |
| This PR  jit.trace | Good | Good |

The reason behind it is the implementation changing. Inspecting to the script (via `unzip`)

```
script_f1.9.0+cu111
├── code
│   ├── __torch__
│   │   └── torch
│   │       └── autograd
│   │           ├── profiler.py   <---- this is the compiled version of the original python file
│   │           └── profiler.py.debug_pkl
│   ├── __torch__.py
│   └── __torch__.py.debug_pkl
├── constants.pkl
├── data.pkl
└── version
```

------

The above conclusion is the result of
```py
import torch
from torch.profiler import record_function

def f(x, y):
    with record_function("r") as rf:
        return x + y

trace_f = torch.jit.trace(f, example_inputs=[torch.ones([1]), torch.ones([1])])
script_f = torch.jit.script(f)

torch.jit.save(trace_f, "trace_f" + torch.__version__ + ".pt")
torch.jit.save(script_f, "script_f" + torch.__version__ + ".pt")
```

```py
func = torch.jit.load(sys.argv[1])
print(func(torch.zeros([1]), torch.ones([1])))
```
